File: Array/yangxin/第三周练习6_15-6_19/20.有效的括号/Solution.java
package com.example.admin.myapplication.stack;

import org.junit.Test;

import java.util.Stack;

/**
 * create by yx
 * on 2020/6/16
 */
public class Solution {
    private Stack<Character> stack = new Stack<>();

    public boolean isValid(String s) {
        if (s == null || s.isEmpty()) {
            return true;
        }

        char[] chars = s.toCharArray();
        for (int i = 0, count = chars.length; i < count; i++) {
            if (stack.isEmpty()) {
                stack.push(chars[i]);
            } else if (stack.peek() == (chars[i] - 2) || stack.peek() == (chars[i] - 1)) {
                stack.pop();
            } else {
                stack.push(chars[i]);
            }
        }
        chars = null;
        return stack.isEmpty();
    }
}

File: Array/yangxin/第三周练习6_15-6_19/20.有效的括号/solution.c
bool isValid(char *s) {
    if (s == NULL || s[0] == '\0') {
        return true;
    }

    int len = strlen(s);
    int index = -1;
    char stack[len];

    for (int i = 0; i < len; i++) {
        if (s[i] == '(' || s[i] == '{' || s[i] == '[') {
            stack[++index] = s[i];
        } else if (index == -1) {
            return false;
        } else if ((stack[index] + 1) == s[i] || (stack[index] + 2) == s[i]) {
            stack[index--] = 0;
        } else {
            stack[++index] = s[i];
        }
    }

    return index == -1;
}